### PR TITLE
fix: exclude empty artifacts from `ContractsByArtifact`

### LINKED ISSUE #7702: Contract verification fails using `forge script` with recent version of forge.
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (63072be 2024-04-18T00:17:28.746005000Z)

### What command(s) is the bug in?

forge script

### Operating System

macOS (Apple Silicon)

### Describe the bug

The following command, run on the default foundry template (forge init) results in the contract being deployed and then subsequently, the verification failing with the following error:

```
forge script script/Counter.s.sol:CounterScript --broadcast --slow --rpc-url base --verify -vvvv
```

```
Submitting verification for [lib/forge-std/src/Base.sol:CommonBase] <address>.
Encountered an error verifying this contract:
Response: `NOTOK`
Details: `Invalid constructor arguments provided. Please verify that they are in ABI-encoded format`
```

It also fails on optimism (so I think it affects all chains, although it is possible it might only affect op-stack chains). 

You will need to add the following to your foundry.toml
```
[rpc_endpoints]
base = "${BASE_RPC_URL}"
optimism = "${OPTIMISM_RPC_URL}"

[etherscan]
base = { key = "${BASE_ETHERSCAN_API_KEY}", url = "${BASE_ETHERSCAN_VERIFIER_URL}" }
optimism = { key = "${OPTIMISM_ETHERSCAN_API_KEY}", url = "${OPTIMISM_ETHERSCAN_VERIFIER_URL}" }
```

I then downgraded to this version of foundry from two weeks ago and the deployment + verification works fine. Note that the eip1559 fix wasn't implemented by this time, so make sure you account for that or you will tip 3gwei. 

https://github.com/foundry-rs/foundry/releases/tag/nightly-f625d0fa7c51e65b4bf1e8f7931cd1c6e2e285e9

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes #7702

## Solution

Exclude artifacts with present but empty bytecode from `ContractsByArtifact`. Such contracts are usually abstract contracts or interfaces. Also adds additional safety check to `VerifyArgs`

### CODE REVIEW

File: crates/script/src/verify.rs [hunk 1]
```diff
@@ -106,6 +106,10 @@ impl VerifyBundle {
         libraries: &[String],
     ) -> Option<VerifyArgs> {
         for (artifact, contract) in self.known_contracts.iter() {
+            // Avoid comparing to empty bytecode.
+            if contract.bytecode.is_empty() {
```
Comment (onbjerg) [inline crates/script/src/verify.rs:110]: we don't need this check if we add the check in `ContractsByArtifact` i think?

File: crates/common/src/contracts.rs [hunk 2]
```diff
@@ -45,6 +45,13 @@ impl ContractsByArtifact {
                     let bytecode = artifact.bytecode.and_then(|b| b.into_bytes())?;
                     let deployed_bytecode =
                         artifact.deployed_bytecode.and_then(|b| b.into_bytes())?;
+
+                    if bytecode.is_empty() {
```
Comment (onbjerg) [inline crates/common/src/contracts.rs:49]: nit: add small comment on what empty bytecode means, just like the pr description

File: crates/script/src/verify.rs (hunk 1 above)
Comment (klkvr) [reply crates/script/src/verify.rs:110]: yeah, but I've seen this bug before so would like to keep this as impact of this bug is pretty high and it's pretty easy to miss while working with artifacts

Comment (klkvr) [reply crates/script/src/verify.rs:110]: also I think would've been caught by verification tests which are disable rn

cc @mattsse

**Review by mattsse — APPROVED** [review]
lgtm 
+ @onbjerg doc nits

**Review by onbjerg — APPROVED** [review]

**Review by DaniPopes — APPROVED** [review]

**Comment by klkvr** [comment]
ah it actually seems that we depend on `known_contracts` to contain ABIs of interfaces and abstract contracts for `targetInterfaces` in invariant tests, looking into fix for this

**Comment by klkvr** [comment]
instead of excluding them we now keep both bytecodes as `Option`s thus we still have info about abstract contracts and interfaces but bytecode is either `None` or correct non-empty bytes

**Pull request merged**